Semicolons terminate statements (#2665)
carbon-language/carbon-lang: proposals/p2665.md

brace, `for ( ... ) { ... }`, or `class C { ...}`.

This does not affect any approved proposal; rather, it makes an important
assumption explicit.

## Problem

Statements need some system for separation. There are two main options for this:

1. Require semicolons to terminate statements.
2. Automatically determine where statements terminate.
    - Some languages, such as Python, define a syntax where a newline terminates
      statements.
    - Other languages, such as Javascript, require semicolons but define rules
      for semicolon insertion.

Although Carbon's design currently assumes semicolons are required, it hasn't
been directly addressed by a proposal.

## Background

### Discussion in Carbon

This was discussed on leads issue
[#1924: Semicolon](https://github.com/carbon-language/carbon-lang/issues/1924).
Some rationale is provided there, stemming from discussion
[#1739: Semicolon](https://github.com/carbon-language/carbon-lang/discussions/1739).

### In other languages

#### Requiring semicolons

In C++, C#, and Java, semicolons are always required.

In Rust, semicolons are generally required, but may be omitted for an

> [implicit return](https://doc.rust-lang.org/std/keyword.return.html).

Reviewer: The Rust rule seems to be a bit [more complicated](https://doc.rust-lang.org/reference/expressions/block-expr.html#:~:text=Statements%20are%20usually%20required%20to%20be%20followed%20by%20a%20semicolon%2C%20with%20two%20exceptions) than that, due to Rust's support of flow-control statements as expressions, and it's worth noting that it leads to [ambiguities](https://doc.rust-lang.org/reference/statements.html#expression-statements).

Author: Added a little more here with the links you noted, does that work for you?

#### Optional semicolons

In Python, a line is a
[simple statement](https://docs.python.org/3/reference/simple_stmts.html), and
parentheses are an idiomatic way to create multi-line statements. Semicolons may
be used to explicitly separate statements. For example:

```python
value = (
  "text"
)
a = 1; b = 2; c = 3
```

Swift takes a different approach, where

> [statement separation is part of the grammar](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/statements).

Reviewer: The Swift documentation suggests that semicolons are simply optional and that the grammar is unambiguous without them (because they can lexically distinguish unary and binary operators), and that they have a separate check that `;` is used when multiple statements appear on the same line. But that doesn't seem to actually be true; Swift's actual parser treats newline and horizontal whitespace differently. For example:
```
// Function call
f (arg1, arg2)

// Uncalled function name statement followed by a separate tuple literal
f
(arg1, arg2)
```
So it seems like Swift is really using a model that's analogous to JavaScript's automatic semicolon insertion, one that treats newlines as statement terminators in some cases, but due to Swift's other language design choices you hit the bad cases much less often in practice -- whereas in JavaScript the common advice is to never rely on automatic semicolon insertion, in Swift the common advice appears to be to always rely on it.

There are certainly cases in Swift where [replacing a space with a newline silently changes semantics](https://swiftfiddle.com/xzatgybsjbcp5ce4ojuugb3pb4), and the actual rules that Swift uses don't seem to be documented anywhere.

Author: I think your first example gives errors:
https://swift.compiler-explorer.com/z/E9EPs71Ez

Your "replacing a space" linked code is not an error, but I think it's distinct for the empty tuple since `make()` is a function call and `()` has no value being discarded. I've added documentation of this as an example of ambiguity.

I would tend to agree that Swift and Kotlin are closer to JavaScript than Python, but I'm wary of the JavaScript comparison for ASI advice. I think there's an important distinction that Swift has better handling of potential mistakes, as demonstrated above, which I think is also at the root of why JavaScript discourages relying on ASI -- it doesn't handle mistakes well.

I've adjusted a little here, but is there more you're looking for, given your comparison?

Reviewer: Yes, my first example gives errors when a newline is present, but that's my point -- the claim made in the Swift grammar that the semicolon is optional and that space and newline are treated as the same kind of whitespace is simply not true. By "statement separation is part of the grammar", I think you mean "the semicolon is simply optional in the grammar, and isn't needed to parse Swift properly", which is the claim that I've seen made in other forums, and that's not right; Swift has some kind of newline-based statement termination rule that doesn't seem to be described in their documentation.

In terms of what I'd want to see here... Swift's actual rule is not written down; what they wrote down isn't what they actually do. There are cases where flowing a statement onto another line at "unlucky" places where whitespace is allowed will change the meaning of the code, but it's not clear how many such constructs exist. Maybe that's what we should be saying here -- Swift's rule seems to work well in practice and their developers seem to like it, but we don't know what it actually is and doesn't match their documentation, so it's hard to assess, and we do know that it has some bad properties in some cases.

Author: I've tried to document the poor documentation of Swift, but I do want to address one detail in your reply:

> By "statement separation is part of the grammar", I think you mean "the semicolon is simply optional in the grammar, and isn't needed to parse Swift properly", which is the claim that I've seen made in other forums, and that's not right

My perspective from flipping through semicolon approaches is there are two fundamental approaches:

1. The language requires semicolons, and there's a way to insert them, as in JavaScript or Go.
2. The language has optional semicolons, and appropriate statement termination is grammatically determined, as in Python, Swift, or Kotlin.

I understand that these two approaches may yield similar results, but I think it's an interesting difference in philosophies. I think Swift is closer to the second. Go and JavaScript are explicitly in the first category. That's what I mean when I say statement separation is part of the grammar.

Note, "properly" sounds subjective to me. While I did mean something different, I actually would be willing to agree the semicolon is optional other than for multi-statement lines. This may just be a different perspective -- while errors can occur with particular code structures, that doesn't mean Swift is wrong for how it treats those code structures.

This allows some statements to wrap lines. However, Swift still disallows
multiple statements on the same line unless a semicolon is present. For example:

```swift
// One statement in Swift, but two in Python and Kotlin.
var x = 1
      + 1
// Two statements in Swift because of whitespace sensitivity.
var x = 1
      +1
```

Kotlin makes

> [newlines usable instead of semicolons in code blocks](https://kotlinlang.org/spec/statements.html#code-blocks).

Reviewer: I think it'd be more useful to say:

Kotlin permits a newline to be used to terminate statements instead of a semicolon. Kotlin's grammar [explicitly enumerates](https://kotlinlang.org/spec/syntax-and-grammar.html) all the places where newlines can appear (see mentions of `NL` in the grammar), and doesn't allow newlines in places where they would introduce ambiguity.
```kotlin
// This is unambiguously parsed as two statements, because
// a newline is not permitted before a `+` operator.
var x = 1
+ 1
```

Author: Done

        as a nuisance to write or visually unnecessary for some developers.
        -   Carbon weighs readability more heavily because of the expectation
            that code will be read more often.
-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
    -   The use of semicolons is expected to improve familiarity for C++
        developers, even for developers who might prefer optional semicolons.

## Alternatives considered

### Optional semicolons

Semicolons could be made optional. This would most likely be with an approach
similar to Python, based mainly on newlines.

Advantages:

-   Languages with optional semicolons are very popular. Python is either the
    most, or the 2nd most, widely used programming language by most measures
    ([1](https://pypl.github.io/PYPL.html)
    [2](https://octoverse.github.com/2022/top-programming-languages)
    [3](https://www.tiobe.com/tiobe-index/)).
-   Echoes the direction of evolution in other languages.
    -   Swift and Kotlin are examples of recently developed languages that make

>         semicolons optional in good ways.

Reviewer: Can we say something more precise than "good" here? Is this "in ways that work well for developers in practice"?

Swift developers seem to like the Swift rule, and it seems hard to find any realistic examples where it does the wrong thing (even though it's not totally clear what rule Swift is using), but I didn't have trouble finding forum posts of Kotlin developers complaining about the Kotlin rule in practice.

Author: Rephrased, but I'm not quickly finding the kotlin rule complaints. What are your search terms?

I did find https://stackoverflow.com/questions/49034127/kotlin-is-ignoring-expression, but the related complaint seems to be a compiler error now.
https://kotlin.compiler-explorer.com/z/5E1GreqjM

Reviewer: Looking a bit harder, it seems like there aren't many complaints. [This](https://discuss.kotlinlang.org/t/there-is-no-total-freedom-for-spacing-in-kotlin/13641) is the one that I found previously, and there's also [this, which isn't really a complaint](https://discuss.kotlinlang.org/t/what-design-principles-contribute-to-the-effectiveness-of-kotlins-asi/15281/2#remaining-problems-2) so much as "I got confused once but I got over it". But most people seem to be saying it works well.

The updated text seems good to me.

Some rationale is provided there, stemming from discussion
[#1739: Semicolon](https://github.com/carbon-language/carbon-lang/discussions/1739).

### In other languages

[This blog](https://pling.jondgoodwin.com/post/semicolon-inference/) provides a
similar survey of multiple languages.

#### Requiring semicolons

In C++, C#, and Java, semicolons are always required.

In Rust, semicolons are generally required, but may be omitted for an
[implicit return](https://doc.rust-lang.org/std/keyword.return.html). Because
[blocks are expressions](https://doc.rust-lang.org/reference/expressions/block-expr.html),
there are
[ambiguities in expression statements](https://doc.rust-lang.org/reference/statements.html#expression-statements)
between parsing as a standalone statement and parsing as part of an expression.

#### Optional semicolons

In Python, a line is a
[simple statement](https://docs.python.org/3/reference/simple_stmts.html), and
parentheses are an idiomatic way to create multi-line statements. Semicolons may
be used to explicitly separate statements. For example:

```python
value = (
  "text"
)
a = 1; b = 2; c = 3
```

> Swift takes a different approach, where
> [statement separation is part of the grammar](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/statements).
> This allows some statements to wrap lines, although Swift still disallows
> multiple statements on the same line (`x = 1 x = 1`) unless a semicolon is

Reviewer: ```suggestion
Swift allows some statements to wrap lines, although it still disallows
multiple statements on the same line (`x = 1 x = 1`) unless a semicolon is
```

Author: Adopted with some edits
